flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerHeap.java

```diff
@@ -0,0 +1,511 @@
+
+		TimerHeapInternalTimer<K, N> currentTimer = heap[idx];
+		int firstChildIdx = idx << 1;
+		int secondChildIdx = firstChildIdx + 1;
+
+		if (secondChildIdx <= heapSize && COMPARATOR.compare(heap[secondChildIdx], heap[firstChildIdx]) < 0) {
+			firstChildIdx = secondChildIdx;
+		}
+
+		while (firstChildIdx <= heapSize && COMPARATOR.compare(heap[firstChildIdx], currentTimer) < 0) {
+			moveElementToIdx(heap[firstChildIdx], idx);
+			idx = firstChildIdx;
+			firstChildIdx = idx << 1;
+			secondChildIdx = firstChildIdx + 1;
+
+			if (secondChildIdx <= heapSize && COMPARATOR.compare(heap[secondChildIdx], heap[firstChildIdx]) < 0) {
+				firstChildIdx = secondChildIdx;
+			}
+		}
+
+		moveElementToIdx(currentTimer, idx);
+	}
+
+	/**
+	 * Returns an unmodifiable set of all timers in the given key-group.
+	 */
+	Set<InternalTimer<K, N>> getTimersForKeyGroup(@Nonnegative int keyGroupIdx) {
+		return Collections.unmodifiableSet(getDedupMapForKeyGroup(keyGroupIdx).keySet());
+	}
+
+	private HashMap<TimerHeapInternalTimer<K, N>, TimerHeapInternalTimer<K, N>> getDedupMapForKeyGroup(
+		@Nonnegative int keyGroupIdx) {
+		return deduplicationMapsByKeyGroup[globalKeyGroupToLocalIndex(keyGroupIdx)];
+	}
+
+	@VisibleForTesting
+	@SuppressWarnings("unchecked")
+	List<Set<InternalTimer<K, N>>> getTimersByKeyGroup() {
+		List<Set<InternalTimer<K, N>>> result = new ArrayList<>(deduplicationMapsByKeyGroup.length);
+		for (int i = 0; i < deduplicationMapsByKeyGroup.length; ++i) {
+			result.add(i, Collections.unmodifiableSet(deduplicationMapsByKeyGroup[i].keySet()));
+		}
+		return result;
+	}
+
+	private void moveElementToIdx(TimerHeapInternalTimer<K, N> element, int idx) {
+		queue[idx] = element;
+		element.setTimerHeapIndex(idx);
+	}
+
+	private HashMap<TimerHeapInternalTimer<K, N>, TimerHeapInternalTimer<K, N>> getDedupMapForKeyGroup(
+		TimerHeapInternalTimer<?, ?> timer) {
+		int keyGroup = KeyGroupRangeAssignment.assignToKeyGroup(timer.getKey(), totalNumberOfKeyGroups);
+		return getDedupMapForKeyGroup(keyGroup);
+	}
+
+	private int globalKeyGroupToLocalIndex(int keyGroup) {
+		checkArgument(keyGroupRange.contains(keyGroup));
+		return keyGroup - keyGroupRange.getStartKeyGroup();
+	}
+
+	private void growIfRequired(int requiredSize) {
+		int oldArraySize = queue.length;
+
+		if (requiredSize >= oldArraySize) {
+			final int grow = (oldArraySize < 64) ? oldArraySize + 2 : oldArraySize >> 1;
+			resizeQueueArray(oldArraySize + grow);
+		}
+		// TODO implement shrinking as well?
+	}
+
+	private void resizeForBulkLoad(int maxTotalSize) {
+		if (maxTotalSize > queue.length) {
+			resizeQueueArray(maxTotalSize + (maxTotalSize >>> 3));
+		}
+	}
+
+	private void resizeQueueArray(int newArraySize) {
+		if (newArraySize - MAX_ARRAY_SIZE > 0) {
+			if (newArraySize < 0 || newArraySize > MAX_ARRAY_SIZE) {
+				throw new OutOfMemoryError("Required timer heap exceeds maximum size!");
+			} else {
+				newArraySize = MAX_ARRAY_SIZE;
+			}
+		}
```

Comment:
Against what are we guarding here exactly? That `newArraySize` is not greater than `MAX_ARRAY_SIZE` and also not negative? Shouldn't it be enough to say `assert(newArraySize > 0 && newArraySize < MAX_ARRAY_SIZE)`?

Reply:
`newArraySize` can become negative via `final int grow = (oldArraySize < 64) ? oldArraySize + 2 : oldArraySize >> 1;`

Reply:
Agreed, but wouldn't it make sense to guard it without the special cases? Simply checking whether `newArraySize` is positive? This would make it easier to understand as well.